client/my-sites/domains/domain-management/settings/cards/domain-redirect-card.tsx

```diff
@@ -2,55 +2,104 @@ import { Button, FormInputValidation } from '@automattic/components';
+	const translate = useTranslate();
+	const { data: redirect, isLoading, isError, error } = useDomainRedirectQuery( domainName );
+
+	// Display success notices when the redirect is updated
+	const { updateDomainRedirect } = useUpdateDomainRedirectMutation( domainName, {
+		onSuccess() {
+			dispatch(
+				successNotice( translate( 'Domain redirect updated and enabled.' ), noticeOptions )
+			);
+		},
+		onError( error: DomainsApiError ) {
+			dispatch(
+				errorNotice(
+					error.message ?? translate( 'An error occurred while updating the redirect.' ),
+					noticeOptions
+				)
+			);
+		},
+	} );
+
+	// Display success notices when the redirect is deleted
+	const { deleteDomainRedirect } = useDeleteDomainRedirectMutation( domainName, {
+		onSuccess() {
+			dispatch(
+				successNotice( translate( 'Domain redirect deleted successfully.' ), noticeOptions )
+			);
+		},
+		onError( error: DomainsApiError ) {
+			dispatch(
+				errorNotice(
+					error.message ?? translate( 'An error occurred while deleting the redirect.' ),
+					noticeOptions
+				)
+			);
+		},
+	} );
+
+	// Render an error if the redirect fails to load
+	useEffect( () => {
+		if ( isError ) {
+			dispatch(
+				errorNotice(
+					( error as DomainsApiError )?.message ??
+						translate( 'An error occurred while fetching your domain redirects.' ),
+					noticeOptions
+				)
+			);
+		}
+	}, [ isError, error, dispatch, translate ] );
 
-const DomainRedirectCard = ( props: DomainRedirectCardProps & PropsFromRedux ) => {
-	const { redirect, target } = props;
-	const { isUpdating, isFetching, isFetched } = redirect;
-	const [ targetUrl, setTargetUrl ] = useState( target ?? '' );
-	const [ protocol, setProtocol ] = useState( redirect.isSecure ? 'https' : 'http' );
-	const { domainName, fetchDomainRedirect, closeDomainRedirectNotice } = props;
+	// Manage local state for target url and protocol as we split redirect target into host, path and protocol when we store it
+	const [ targetUrl, setTargetUrl ] = useState( '' );
+	const [ protocol, setProtocol ] = useState( 'https' );
 	const [ isValidUrl, setIsValidUrl ] = useState( true );
 
+	// Load saved redirect into local state
 	useEffect( () => {
-		fetchDomainRedirect( domainName );
-		return () => {
-			closeDomainRedirectNotice( domainName );
-		};
-	}, [ domainName, fetchDomainRedirect, closeDomainRedirectNotice ] );
+		if ( isLoading || ! redirect ) {
+			return;
+		}
 
-	useEffect( () => {
-		if ( isFetched && ! isUpdating ) {
-			setTargetUrl( target ?? '' );
-			setProtocol( redirect.isSecure ? 'https' : 'http' );
+		try {
+			const path = ( redirect.targetPath ?? '' ) as string;
+			const origin =
+				( redirect.isSecure ? 'http://' : 'https://' ) +
+				( redirect.targetHost ?? '_invalid_.domain' );
+			const url = new URL( path, origin );
+			if ( url.hostname !== '_invalid_.domain' ) {
+				setTargetUrl( url.hostname + url.pathname + url.search + url.hash );
+				setProtocol( redirect.isSecure ? 'https' : 'http' );
```

Comment:
If `targetUrl` and `protocol` are derived from query data it may be worth exploring whether the `select` option provided is a better fit ([explanation](https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option), [docs](https://tanstack.com/query/v4/docs/react/reference/useQuery)). I didn't double-check so it may not applicable but it's a useful option to know.